### PR TITLE
Fixes #3281 - Convert issue page to be server rendered (part 1)

### LINKED ISSUE #3281: Convert issue page to be server rendered
It should be possible to render an issue from the server, since we get all data from GitHub.

New comments will continue to be dynamic. 

See https://github.com/webcompat/webcompat.com/issues/2449#issuecomment-611619573

### PR DESCRIPTION
**Pull request by miketaylr**
Right now I'm going to leave this as a draft. I decided to not tackle existing comments here just to keep things small. I'll file a follow-up on that and tackle it once this lands.

### CODE REVIEW

File: webcompat/views.py [hunk 1]
```diff
@@ -349,7 +350,11 @@ def show_issue(number):
     if session.get('show_thanks'):
         flash(number, 'thanks')
         session.pop('show_thanks')
-    return render_template('issue.html', number=number)
+    # right now this is a double request... can we prevent
+    # that from happening? add in a caching layer? (maybe it already happens)
+    issue_data = proxy_issue(number)
```
Comment (miketaylr) [inline webcompat/views.py:355]: I think I can prevent the client from hitting our API endpoint entirely for an Issue (except for fetching labels, but we should be able to do that via the server eventually I think).

File: webcompat/helpers.py [hunk 2]
```diff
@@ -444,6 +444,13 @@ def wrapped_func(*args, **kwargs):
                 file_path = FIXTURES_PATH + request.path
             if not file_path.endswith('.json'):
                 file_path = file_path + '.json'
+            if not os.path.exists(file_path):
```
Comment (miketaylr) [inline webcompat/helpers.py:447]: If I can solve the client XHR request, we can remove this hack (I'm pretty sure).

**Comment by miketaylr** [comment]
To investigate (how I broke this)

```
× firefox on linux 4.15.0-1052-aws - Issues - Simple title concatenation (0.46s)
    AssertionError: The title has been concatenated.: expected 'www.prudential.com.sg - design is broken' to equal 'www.prudential.com.sg - Text rendering missing'
    
    A www.prudential.com.sg - design is broken
    E www.prudential.com.sg - Text rendering missing
    
      at Command.<anonymous> @ tests/functional/issues-non-auth.js:124:18
      at Test.Simple title concatenation [as test] @ tests/functional/issues-non-auth.js:123:10
      @ src/lib/Test.ts:267:51
```

**Comment by miketaylr** [comment]
> To investigate (how I broke this)

Looks like I need to port `getTitle` from the issue.js model. Can probably make a jinja filter to do the heavy lifting.

File: webcompat/templates/__init__.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+
+@app.template_filter('format_date')
+def format_date(datestring):
+    """For now, just chops off crap."""
+    # 2014-05-01T02:26:28Z
+    return datestring[0:10]
+
+
+@app.template_filter('format_title')
```
Comment (miketaylr) [inline webcompat/templates/__init__.py:60]: Here's the source I was porting from https://github.com/webcompat/webcompat.com/blob/c34522f95a14ed6efa903a40e1fc8902c2d2b11e/webcompat/static/js/lib/models/issue.js#L73

It's mostly the same, except for `get_description` -- since it was running in the context of JS / DOM, it was simpler to grab text w/o tags (via `textConent` to run the regexp on. Luckily, flask provides a `striptags` method: https://flask.palletsprojects.com/en/1.1.x/api/#flask.Markup.striptags

**Comment by miketaylr** [comment]
progress!

```
× firefox on linux 4.15.0-1052-aws - Milestones (non-auth) - Page loads without milestone set (0.17s)
    AssertionError: expected 'No - see issue description' to include 'No Milestone.'
      at Command.<anonymous> @ tests/functional/milestones-non-auth.js:27:18
      at Test.Page loads without milestone set [as test] @ tests/functional/milestones-non-auth.js:25:10
      @ src/lib/Test.ts:267:51
```

File: webcompat/templates/__init__.py (hunk 3 above)
Comment (karlcow) [reply webcompat/templates/__init__.py:60]: I'm closely following what you are doing these days and loving it.

**Comment by miketaylr** [comment]
TODO:

* [ ] create a template filter to handle the milestone / state text (Closed: Needs Diagnosis, or Needs Diagnosis, etc.) Right now we just show the milestone name (which is a single lowercase word)
* [ ] fix bugs in `AsideView.render()` -- I forgot about handling the state (closed vs open)
* [ ] Remove the xhr request to the API endpoint (if we can get away with it, I haven't thought much about it... might require dumping initial state for asideview into the document)

**Comment by miketaylr** [comment]
Getting closer. Need to investigate this local error and add some error handling that I deleted from the client.

```
___________________________________________________________________________________________ TestHTTPCaching.test_not_modified_status ___________________________________________________________________________________________

self = <unit.test_http_caching.TestHTTPCaching testMethod=test_not_modified_status>

    def test_not_modified_status(self):
        '''Checks if we receive a 304 Not Modified.'''
        for uri in ['/about',
                    '/contributors',
                    '/issues',
                    '/issues/100',
                    '/privacy']:
            rv = self.app.get(uri, environ_base=html_headers)
            etag = rv.get_etag()[0]
            rv2 = self.app.get(uri,
                               environ_base=html_headers,
                               headers={'If-None-Match': etag})
>           self.assertEqual(rv2.status_code, 304)
E           AssertionError: 200 != 304

tests/unit/test_http_caching.py:59: AssertionError
```

File: webcompat/templates/issue.html [hunk 4]
```diff
@@ -42,6 +45,9 @@
   </main>
 {% endblock %}
 {% block extrascripts %}
+<script nonce={{nonce}}>
+  var issueData = {{ issue_data|tojson }};
```
Comment (miketaylr) [inline webcompat/templates/issue.html:49]: So uh, the lazier approach would have been to do this, and bootstrap the existing Issue model from this and call it a day. Archived issues would just have the response hard-coded, and the client would render them.

After review, if we're not happy with the mess I'm making here, I'm happy to throw this all away and submit another PR from this POV. :)

File: webcompat/views.py (hunk 1 above)
Comment (miketaylr) [reply webcompat/views.py:348]: Probably we should try https://github.com/webcompat/webcompat.com/blob/master/webcompat/helpers.py#L498 to not break HTTP caching.

**Comment by miketaylr** [comment]
> edit: I'm guessing that since we don't hit the API endpoint, we don't get an etag?

No, it's simpler than that:

```
🐓 diff ~/Desktop/a.html ~/Desktop/b.html 
493c493
< <script nonce=fb8398749dec1b6f449308e3b36e7e29f7f6cf1d>
---
> <script nonce=12e11dfe51b78df09c2569023c488fcd914ed388>
```

We're injecting this as JS, and we need a `nonce` to get around our CSP (which is always different...so we generate a new etag). There's a few options to get around this, I think.

**Comment by miketaylr** [comment]
OK, getting closer. HTTP caching fixed. Just need to consider the error from GitHub and how to handle that. Then this should be ready for a review (and then on to other improvements).

File: tests/fixtures/issues/2.json [hunk 5]
```diff
@@ -0,0 +1 @@
+webcompat.com/tests/fixtures/api/issues/2.json
```
Comment (miketaylr) [inline tests/fixtures/issues/2.json:1]: TODO: remove this, I'm not bothering with symlinks anymore.

File: tests/unit/test_urls.py [hunk 6]
```diff
@@ -206,16 +206,16 @@ def test_issue_int(self):
         * an issue only displays if <number> is an integer
         * /issues/<number> exists, and does not redirect.
         """
-        rv = self.app.get('/issues/3')
+        rv = self.app.get('/issues/2')
```
Comment (miketaylr) [inline tests/unit/test_urls.py:209]: (this commit can be reverted too)

File: webcompat/helpers.py (hunk 2 above)
Comment (miketaylr) [reply webcompat/helpers.py:413]: I decided to live with this hack, it's just mocking test code. The alternative is to move around a bunch of fixture files, and it doesn't seem worth the effort. 🤷

File: tests/unit/test_urls.py (hunk 6 above)
Comment (miketaylr) [reply tests/unit/test_urls.py:209]: (actually no, it needs a fixture)

**Comment by miketaylr** [comment]
OK, ready for review now. Not urgent, any time this week. 🙇 🙇‍♀️ 

r? @ksy36 for JS
r? @karlcow for Python

File: webcompat/templates/__init__.py [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        'duplicate': 'Duplicate',
+        'incomplete': 'Incomplete',
+        'fixed': 'Fixed',
+        'invalid': 'Invalid',
+        'needscontact': 'Needs Contact',
+        'needsdiagnosis': 'Needs Diagnosis',
+        'needstriage': 'Needs Triage',
+        'non-compat': 'Non-Compat',
+        'sitewait': 'Site Contacted',
+        'wontfix': 'Wont Fix',
```
Comment (karlcow) [inline webcompat/templates/__init__.py:114]: nit: should it be `Won't Fix`?

File: webcompat/views.py [hunk 8]
```diff
@@ -41,14 +42,14 @@
 from webcompat.helpers import is_valid_issue_form
 from webcompat.helpers import prepare_form
 from webcompat.helpers import set_referer
-from webcompat.issues import report_issue
 from webcompat.helpers import get_extra_labels
 from webcompat.helpers import mockable_response
+from webcompat.issues import report_issue
+from webcompat.templates import bust_cache
```
Comment (karlcow) [inline webcompat/views.py:48]: nit: this one is below. ;)

File: webcompat/views.py [hunk 9]
```diff
@@ -349,7 +343,10 @@ def show_issue(number):
     if session.get('show_thanks'):
         flash(number, 'thanks')
         session.pop('show_thanks')
-    return render_template('issue.html', number=number)
+    issue_data = proxy_issue(number)
+    return render_template('issue.html',
+                           issue_data=json.loads(issue_data[0]),
+                           json_data=issue_data[0].decode('utf-8'))
```
Comment (karlcow) [inline webcompat/views.py:349]: There are two methods to convert data-types :) 
https://github.com/webcompat/webcompat.com/blob/ccedceb3d38c07b9a6611b8ebed57553ac1b76e0/webcompat/helpers.py#L702-L717

```suggestion
                           json_data=helpers.to_str(issue_data[0])
```

File: tests/unit/test_template.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+    def setUp(self):
+        """Set up."""
+        pass
+
+    def tearDown(self):
+        """Tear down."""
+        pass
+
+    def test_format_date(self):
+        """We trust what GitHub gives us, so just test length."""
```
[the remark this answers is not in the document]
Comment (miketaylr) [reply tests/unit/test_template.py:31]: Great suggestion thanks.

File: webcompat/helpers.py [hunk 11]
```diff
@@ -444,12 +410,18 @@ def wrapped_func(*args, **kwargs):
...
+                # doesn't exist (it's in /api/)-- so rather than duplicate
+                # files, we just look there.
+                file_path = FIXTURES_PATH + '/api' + request.path + '.json'
             if not os.path.exists(file_path):
                 print('Fixture expected at: {fix}'.format(fix=file_path))
                 print('by the http request: {req}'.format(req=request.url))
                 return ('', 404)
             else:
-                with open(file_path, 'r') as f:
+                with open(file_path, 'rb') as f:
```
[the remark this answers is not in the document]
Comment (miketaylr) [reply webcompat/helpers.py:424]: No problem. It seems like it didn't matter (until now).

File: webcompat/templates/__init__.py (hunk 7 above)
Comment (miketaylr) [reply webcompat/templates/__init__.py:114]: Nice catch!

File: webcompat/views.py (hunk 9 above)
Comment (miketaylr) [reply webcompat/views.py:349]: Oh! I didn't notice these. Perfect, thanks.

File: webcompat/views.py (hunk 8 above)
Comment (miketaylr) [reply webcompat/views.py:48]: (@karlcow -- sorry, I don't follow. do you mean `from webcompat.templates import bust_cache` should be moved down below?)

Comment (karlcow) [reply webcompat/views.py:48]: yes sorry that was unclear, though at the same time, I see that the order is a bit messed up for other things. 

Let's do that later.

**Comment by miketaylr** [comment]
Heh, managed to mess things up while rebase. `git reflog` to the rescue...

**Comment by miketaylr** [comment]
OK, apologies for the messy merge commit. 

r? @karlcow

File: tests/unit/test_template.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+
+    def test_format_milestone_class(self):
+        """Test that we return the correct class string."""
+        issue_data = {}
+        issue_data['milestone'] = {}
+        issue_data['milestone']['title'] = 'needsdiagnosis'
+        issue_data['state'] = 'open'
+        self.assertEqual(format_milestone_class(issue_data), 'needsdiagnosis')
+        issue_data['state'] = 'closed'
+        self.assertEqual(format_milestone_class(issue_data), 'closed')
```
Comment (karlcow) [inline tests/unit/test_template.py:105]: same logic than previous comment 

```suggestion
        # open state
        issue_data = {'milestone': {'title': 'needsdiagnosis'}, 'state': 'open'}
        assert format_milestone_class(issue_data) == 'needsdiagnosis'
        # closed state
        issue_data = {'milestone': {'title': 'needsdiagnosis'}, 'state': 'closed'}
        assert format_milestone_class(issue_data) == 'closed'
```

File: tests/unit/test_template.py [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+<strong>Description</strong>: Desc from Description<br>
+<strong>Steps to Reproduce</strong>:<br>
+"""
+        formatted_title = format_title(issue_data)
+        self.assertEqual(formatted_title,
+                         'staging.webcompat.com - Desc from Description')
+        issue_data['body_html'] = 'fix my bug'
+        formatted_title = format_title(issue_data)
+        self.assertEqual(formatted_title,
+                         'staging.webcompat.com - desc from title')
```
Comment (karlcow) [inline tests/unit/test_template.py:80]:
```suggestion
        # title comes from body description
        issue_data = {
            'body_html': '\n'
            '<p><strong>Problem type</strong>: Site is not usable<br>\n'
            '<strong>Description</strong>: Desc from Description<br>\n'
            '<strong>Steps to Reproduce</strong>:<br>\n',
            'title': 'staging.webcompat.com - desc from title'}
        expected = 'staging.webcompat.com - Desc from Description'
        assert format_title(issue_data) == expected
        # title comes from issue title
        issue_data = {
            'body_html': 'fix my bug'
            'title': 'staging.webcompat.com - desc from title'}
        expected = 'staging.webcompat.com - desc from title'
        assert format_title(issue_data) == expected
```

**Comment by karlcow** [comment]
some edits, one rebase, let's run the tests.

**Comment by karlcow** [comment]
And this is merged.

**Comment by miketaylr** [comment]
❤️

**Comment by magsout** [comment]
Super nice work 👍
